MudChart: Timeseries charting improvements

radderz opened the pull request:
Improve the X Axis labelling for time series

## Description
The Timeseries before the change resulted in the first time label on the chart being the min timestamp of the data and it didn't round correctly to the label spacing option, depending on the format used this could lead to having the wrong values shown.

In this PR I have introduced three new settings on the timeseries chart "TimeLabelSpacingRounding", "TimeLabelSpacingRoundingPadSeries" and "MatchBoundsToSize". 

The first option "TimeLabelSpacingRounding" will make sure the labels are on "ticks" of the Label spacing. This means that if the first datetime on the chart is not directly on a tick the label will be offset. The second option "TimeLabelSpacingRoundingPadSeries" will instead offset the series instead of the labels so the first label will be at the 0 point of the X Axis and the chart will start there or after depending on the first datapoint. "TimeLabelSpacingRoundingPadSeries" will only be respected if "TimeLabelSpacingRounding" is true.

"MatchBoundsToSize" will fill the chart to the size provided, this only works if using "px" sizes for both width and height.

## How Has This Been Tested?
visually and I have added unit tests, and also added the basic unit test that was missing from my original PR that added the time series chart.

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

Original (and still default)
![image](https://github.com/user-attachments/assets/448ead60-ab1a-4f74-b67e-93bc23ff345c)

TimeLabelSpacingRounding = true.
![image](https://github.com/user-attachments/assets/1342ad65-293a-4ee6-b4ac-34cbd15700cf)

TimeLabelSpacingRoundingPadSeries = true.
![image](https://github.com/user-attachments/assets/ae13e5ef-b624-48e6-a7b8-5b9736397d9f)

MatchBoundsToSize = true
![image](https://github.com/user-attachments/assets/447ea328-96a5-41c4-88cb-a29d3a4833c1)


## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

radderz committed:
Fix timestamps of x axis for when the data starts/ends outside of a tick, before the timestamps of the ticks would start from the first datapoint, which could be anywhere within a tick, by doing this the positioning of the chart correctly sits within the desired ticks .

radderz committed:
addunit tests for timeseries and new features

radderz commented:
No idea what those ValidAttributeTests failures are about but is unrelated to what I have changed in this PR

radderz committed:
Add custom bounds sizing for time series charts based on the size of the parent.

radderz commented:
Added a way to make the chart fill the size

radderz committed:
clear whitespace

ScarletKuro commented:
> No idea what those ValidAttributeTests failures are about but is unrelated to what I have changed in this PR

Currently some tests are broken due to the new 9.0.200 SDK

radderz commented:
@ScarletKuro the other approach I thought about was to use JavaScript interop with a size observer to allow for any form of size based logic on the outer chart. I will in a different PR do the other charts that make sense to support filling the parent (bar/stacked bar)

radderz committed:
support bar, stacked bar, line bounds changing
support stacked bar width ratio and rewrite in general to be more similar to the other charts, tweak for better UX.

radderz committed:
move common code to single place, add AxisOptions model for doing label rotation, extra height for labels and label step.

ScarletKuro committed:
Merge branch 'dev' into charting-improvements-2025-02

ScarletKuro commented on src/MudBlazor/Components/Chart/Charts/TimeSeries.razor.cs:
```diff
@@ -14,8 +14,10 @@ namespace MudBlazor.Charts
-        private const double BoundHeight = 350.0;
+        private const double BoundWidthDefault = 800;
+        private const double BoundHeightDefault = 350;
+        private double BoundWidth = BoundWidthDefault;
```
private non constant fields in our code style are `_xyzField`.

radderz replied:
private consts in the editorconfig actually complain if doing the _xyzField style so I have stuck with the upper case, but the private double BoundWidth should be _xyzField style.

I am doing another push with all the other changes to charts I have done shortly.

ScarletKuro replied:
as I said **non** constant fields.
const fields with upper letter, that's correct, however for example:
` private double BoundWidth` is not a constant, therefore `private double _boundWidth`

ScarletKuro commented on src/MudBlazor.UnitTests/Components/Charts/TimeSeriesChartTests.cs:
```diff
@@ -11,12 +12,109 @@ namespace MudBlazor.UnitTests.Charts
+            comp.Markup.Should().BeEquivalentTo(TimeSeriesMarkupBasic);
+        }
+
+
```
Remove extra new line

radderz committed:
stacked bar chart hover working

radderz committed:
add area feature for line chart

radderz committed:
merge in category upgrades

radderz committed:
add in auto chart sizing based on element size
standardise for all x/y charts like line, bar, stacked bar, timeseries
simplify the settings locations to be consistent between the different charts
add data labels as a feature
add element size watching helper function
add svg bbox function for sizing the box to match the inside text in the svg datapoint label popup
update examples to have more options to showcase new features
fix spline X axis being wrong spacing due to missing -1 on division.
support area line charts
add a rounding on the charts as there is no reason to have high precision subpixels which bloats out the svg content
update all tests to consider the rounding and other changes like labels

radderz committed:
Merge branch 'charting-improvements-2025-02' of https://github.com/radderz/MudBlazor into charting-improvements-2025-02

radderz committed:
reorder scheduledCall = null

radderz commented:
#Summary

I have added a bunch more enhancements and fixes to charts rather than do a second PR:

- add in auto chart sizing based on element size
- standardise for all x/y charts like line, bar, stacked bar, timeseries
- simplify the settings locations to be consistent between the different charts
- add data labels as a feature
- add element size watching helper function
- add svg bbox function for sizing the box to match the inside text in the svg datapoint label popup
- update examples to have more options to showcase new features
- fix spline X axis being wrong spacing due to missing -1 on division.
- support area line charts
- add a rounding on the charts as there is no reason to have high precision subpixels which bloats out the svg content
- update all tests to consider the rounding and other changes like labels

# Bar Chart

Can match the size to the bounds, rotate the labels:
![image](https://github.com/user-attachments/assets/e31eb587-f676-4983-ba47-de373597185f)

# Stacked Bar Chart
Can match the size to the bounds, rotate the labels, change the bar thickness, has a slight overlap of the bars to remove the floating point rounding causing lines between bar sections:
![image](https://github.com/user-attachments/assets/43f1572c-a915-4e56-8ff9-fda2f108f645)

# Line Chart
Can match the size to the bounds, rotate the labels, set lines to area mode, show data markers always and on hover of a data marker it will show a label. The data markers are always there, if show data markers is off they are hidden until hovered over.
![image](https://github.com/user-attachments/assets/ed315389-f5a6-422e-9088-ec4865b92320)
![image](https://github.com/user-attachments/assets/fae42611-dac0-4e4d-99f2-8edb89fdc428)
![image](https://github.com/user-attachments/assets/200ca860-4dd0-4b54-a601-f4ad0b6f3840)

Also supports interpolated mode (both area and markers)
![image](https://github.com/user-attachments/assets/1d4b0dc6-0551-44c1-af33-983bee5371d4)

# Timeseries Chart
Can match the size to the bounds (with , rotate the labels, set lines to area mode, show data markers always and on hover of a data marker it will show a label. The data markers are always there, if show data markers is off they are hidden until hovered over.
![image](https://github.com/user-attachments/assets/8ed48363-a686-4d67-ac52-1219472ee121)

There is a new feature for how the x axis is rendered as originally I made it work such that the first label will be the earliest data, but if you want the labels to be rounded to the time label spacing which is probably most of time then there is two new properties on the chart. "TimeLabelSpacingRounding" and "TimeLabelSpacingRoundingPadSeries" as explained in the original post.

# Data Marker Tooltip

The tooltip supports 3 tags and two lines (Title and Subtitle). {{X_VALUE}}, {{Y_VALUE}}, {{SERIES_NAME}}. The timeseries has an additional property on the MudTimeSeriesChart called DataMarkerTooltipTimeLabelFormat which allows for setting what the Datetime format for the X_VALUE should be done as.

![image](https://github.com/user-attachments/assets/d0f4ff48-9849-4df2-8cf3-a1e7574c253b)

radderz committed:
fix some issues with sizing

radderz committed:
fix test

radderz committed:
fix whitespacing

radderz committed:
undo change to docs.csproj

Anu6is commented:
Definitely a worthwhile revamp of the chart components! 👍 

Looks like you've added the tooltip to most of the charts—would you mind adding it to the Pie and Donut charts as well? That would help ensure consistency across all chart types.

radderz committed:
add hover colour change to everything except bar/stackbar as there is an issue with the filter logic
add data tooltips to bar/donut/pie charts
update tests

radderz committed:
remove double tooltip

radderz committed:
change tooltip to show data value instead of percentage

radderz commented:
I have added to pie, donut and bar, I had to rewrite the pie/bar to be able to have a reasonably sized tooltip.

Pie Chart
![image](https://github.com/user-attachments/assets/08d6d36f-e577-45be-b5e7-3f4462f45c12)

Donut Chart
![image](https://github.com/user-attachments/assets/0b13db50-ddb8-4dd1-b9fa-8f20bd76ff8c)

Bar Chart
![image](https://github.com/user-attachments/assets/ad1d739b-56c4-42f5-9fa1-b3e12f74c736)

radderz commented:
Have to stop working any further on this (other than some minor bug fixes/code style fixes) as I have burned way more time than expected!

radderz committed:
fix positioning of the tooltip to be the middle of the segment, it doesn't need the stroke sizing removed

radderz commented:
@Anu6is @ScarletKuro is this PR as it is good enough now to complete?

ScarletKuro commented on src/MudBlazor/Components/Chart/Charts/Pie.razor:
```diff
@@ -1,13 +1,41 @@
+                  d="@item.Data"
+                  @onmouseover="(e) => OnSegmentMouseOver(e, item)"
+                  @onmouseout="OnSegmentMouseOut">
+              </path>
```
Wrong indent

ScarletKuro commented on src/MudBlazor/Components/Chart/Models/LineDisplayType.cs:
```diff
@@ -1,7 +1,7 @@
 namespace MudBlazor;
 
-public enum TimeSeriesDisplayType
+public enum LineDisplayType
```
This a breaking change, we cannot do such renaming.
Unless we just move this to v9.

radderz replied:
I have made it backwards compatible, I readdd the timeseriesdisplaytype and added an obsolete, on the timeseries that actually had that enum, I added a getter/setter to set the same new property.

ScarletKuro reviewed (changes requested):
There are 47 new quality issues: https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&sinceLeakPeriod=true&pullRequest=10865&id=MudBlazor_MudBlazor
Please, fix those.

radderz committed:
code style updates

radderz committed:
Deprecate the donut as its just a pie chart with a path with an inner arc instead of going all the way back to the center point

radderz commented:
I have done the recommended changes

I have rewritten the pie chart to handle the donut rendering too as the concept is basically the same except for the hole in the middle, this means less code to maintain going forward.

radderz committed:
fix indentation

radderz committed:
change exception type

radderz committed:
some more minor tweaks

radderz committed:
merge in dev

radderz committed:
ensure the range rather than exception

radderz committed:
Revert donut from being obsolete, and instead just wrap the pie chart and pass down the same parameters

radderz committed:
format code

radderz committed:
fix tests

Anu6is commented on src/MudBlazor.UnitTests/Components/Charts/DonutChartTests.cs:
```diff
@@ -40,8 +40,8 @@ public void Init()
         {
-            var comp = Context.RenderComponent<Donut>();
-            comp.Markup.Should().Contain("mud-chart-donut");
+            var comp = Context.RenderComponent<Pie>();
```
Since you reverted obsoleting `Donut` and instead made it wrap `Pie`, shouldn't the test use `Donut`

radderz replied:
Yea this is right, I missed that when I changed it back to make it not a breaking change. That's a quick fix.

radderz replied:
@Anu6is - This has been updated back to using the donut component, I also made it so that it switches it's class to be the same as it was when its a donut.

ScarletKuro commented:
> Are any of the changes breaking?

Well it kinda still is as it removed `MudChartParent` from Donut, Line, Stackbar etc, not sure how much is this used tho.

henon commented:
> > Are any of the changes breaking?
> 
> Well it kinda still is as it removed `MudChartParent` from Donut, Line, Stackbar etc, not sure how much is this used tho.

Then we should merge it in the first v9 preview just to be sure.

radderz committed:
Change donut tests to use donut instead, swapped the class too to be the same as the old one when in donut mode.

Anu6is commented on src/MudBlazor/Components/Chart/Charts/Pie.razor.cs:
```diff
@@ -13,14 +17,22 @@ namespace MudBlazor.Charts
         [CascadingParameter]
         public MudChart? MudChartParent { get; set; }
 
+        // 1 = full circle, 0.25 = donut with 25% radius thickness 75% hole.
```
Use `<summary>` as this is transferred to the documentation page. Also add more of a description

Anu6is commented on src/MudBlazor/Components/Chart/MudChartBase.cs:
```diff
@@ -21,6 +22,13 @@ public abstract class MudChartBase : MudComponentBase
     public ChartOptions ChartOptions { get; set; } = new();
 
+    /// <summary>
+    /// Display options applied to the chart specific to Axis charts.
```
Display options for axis-based charts

Anu6is commented on src/MudBlazor/Components/Chart/MudTimeSeriesChart.razor.cs:
```diff
@@ -19,13 +19,36 @@ public abstract class MudTimeSeriesChartBase : MudChartBase
         public TimeSpan TimeLabelSpacing { get; set; } = TimeSpan.FromMinutes(5);
 
+        /// <summary>
+        /// A way to have the timestamp labels be rounded to the nearest spacing value. Default is false.
```
```css
/// <summary>
/// Determines whether timestamp labels should be rounded to the nearest spacing value. 
/// </summary>
/// <remarks>
/// Default is <c>false</c>.
/// </remarks>
```

Anu6is commented on src/MudBlazor/Components/Chart/MudTimeSeriesChart.razor.cs:
```diff
@@ -19,13 +19,36 @@ public abstract class MudTimeSeriesChartBase : MudChartBase
+        /// When TimeLabelSpacingRounding is true, 
+        /// When true: pad series to allow rounding with labels before and after the series start/end.
+        /// When false: move the labels inwards to be rounded to the label spacing without changing the start/end time of the axis.
+        /// </summary>
```
```css
/// <summary>
/// Determines how timestamp labels are adjusted when <see cref="TimeLabelSpacingRounding"/> is enabled.
/// </summary>
/// <remarks>
/// When <c>true</c>, the series is padded to allow rounding with labels before and after the series start and end.
/// When <c>false</c>, labels are moved inward to align with the label spacing without altering the axis start and end times.
/// </remarks>
```

Anu6is commented on src/MudBlazor/Components/Chart/MudTimeSeriesChart.razor.cs:
```diff
@@ -19,13 +19,36 @@ public abstract class MudTimeSeriesChartBase : MudChartBase
         /// <summary>
         /// A way to specify datetime formats for timestamp labels, default of HH:mm.
         /// </summary>
         [Parameter]
```
```css
/// <summary>
/// Specifies the datetime format for timestamp labels. 
/// </summary>
/// <remarks>
/// Defaults to <c>"HH:mm"</c>.
/// </remarsk>
```

Anu6is commented on src/MudBlazor/Components/Chart/MudTimeSeriesChart.razor.cs:
```diff
@@ -19,13 +19,36 @@ public abstract class MudTimeSeriesChartBase : MudChartBase
 
+        /// <summary>
+        /// A way to specify datetime formats for timestamp labels used in datapoint marker labels, default of HH:mm.
+        /// </summary>
```
```css
/// <summary>
/// Specifies the DateTime format for Timestamp labels in DataPoint marker tooltips. 
/// </summary>
/// <remarks>
/// Defaults to <c>"HH:mm"</c>.
/// </remarks>
```

Anu6is commented on src/MudBlazor/Components/Chart/MudTimeSeriesChart.razor.cs:
```diff
@@ -39,5 +62,13 @@ public abstract class MudTimeSeriesChartBase : MudChartBase
+
+        /// <summary>
+        /// Specifies the chart should take its bounds from the parent chart.
+        /// </summary>
```
```css
/// <summary>
/// Determines if the chart should derive its bounds from the parent chart.
/// </summary>
```

Anu6is commented on src/MudBlazor/Components/Chart/Parts/ChartTooltip.razor.cs:
```diff
@@ -0,0 +1,37 @@
+{
+    public partial class ChartTooltip
+    {
+        [Parameter, EditorRequired] public string Title { get; set; } = string.Empty;
```
missing comments

radderz committed:
Comments changes as requested by Anu6is

radderz commented:
@Anu6is thats done

ScarletKuro commented:
@Anu6is can you look why `Open_CloseBySelectingADate_CheckClosed_Check_DateChangedCount` started to fail randomly? I think you were dealing with this kind of problems with datapicker before.

ScarletKuro committed:
Some nits

ScarletKuro commented on src/MudBlazor/Components/Chart/Charts/TimeSeries.razor.cs:
```diff
@@ -35,6 +47,20 @@ partial class TimeSeries : MudTimeSeriesChartBase
+
+        public TimeSeries()
+        {
+            _dotNetObjectReference = DotNetObjectReference.Create(this);
```
I think `TimeSeries` is missing the Dispose of the  `_dotNetObjectReference`

radderz replied:
Good catch, I added it to the other base class and missed this one, thats pushed.

radderz committed:
add missing dispose for dotnetobjectref

ScarletKuro committed:
Fix formatting

radderz committed:
Merge branch 'charting-improvements-2025-02' of https://github.com/radderz/MudBlazor into charting-improvements-2025-02

ScarletKuro reviewed (approved):
LGTM

ScarletKuro committed:
Fix trimming issues

ScarletKuro committed:
nit: we use JsRuntime instead of JSRuntime

ScarletKuro committed:
Merge branch 'dev' into charting-improvements-2025-02

ScarletKuro commented:
> @Anu6is can you look why `Open_CloseBySelectingADate_CheckClosed_Check_DateChangedCount` started to fail randomly? I think you were dealing with this kind of problems with datapicker before.

Seems to be dependent on culture, passes locally, but it I set:
```
[Test]
[SetCulture("en-US")]
[SetUICulture("en-US")]
public void Open_CloseBySelectingADate_CheckClosed_Check_DateChangedCount()
```
Then it fails

Anu6is commented:
> @Anu6is can you look why `Open_CloseBySelectingADate_CheckClosed_Check_DateChangedCount` started to fail randomly? I think you were dealing with this kind of problems with datapicker before.

Submitted #10961

ScarletKuro committed:
Merge branch 'dev' into charting-improvements-2025-02

radderz committed:
Merge branch 'dev' into charting-improvements-2025-02

henon commented:
Thanks @radderz !
